### ISSUE TITLE
a bug of memory leakage based on tensorflow backend

### ISSUE THREAD

**Issue by guoxiaolu** [keras-team/keras#8495]
`def get_gradients(model, x, layers_name):
    gradients_all = model.optimizer.get_gradients(model.total_loss, weights)  # gradient tensors

    input_tensors = [model.inputs[0],  # input data
                     model.sample_weights[0],  # how much to weight each sample by
                     model.targets[0],  # labels
                     K.learning_phase(),  # train or test mode
                     ]

    get_gradients = K.function(inputs=input_tensors, outputs=gradients_all)
    out = get_gradients(x)
    return dict(zip(layers_name, out))`

`if __name__ == '__main__':

    model = vgg16.VGG16()
    sgd = SGD(lr=1e-2, decay=1e-6, momentum=0.9, nesterov=True)
    model.compile(optimizer=sgd, loss='categorical_crossentropy', metrics=['accuracy'])
    model.summary()
    graph = K.get_session().graph

    # datagenerator
    gen = ImageDataGenerator(preprocessing_function=processing_function)
    train_generator = gen.flow_from_directory(train_img_path, target_size=image_size, classes=classes, shuffle=True,
                                              batch_size=batch_size)

    # Get gradient tensors
    trainable_weights = model.trainable_weights  # weight tensors
    weights = []
    layers_name = []
    # weight name is different from layer name, as the weight is consisted of kernel and bias
    for weight in trainable_weights:
        if model.get_layer(weight.name.split('/')[0]).trainable:
            weights.append(weight)
            layers_name.append(weight.name[:-2])

    total = 0
    gradients = dict(zip(layers_name, [0]*len(layers_name)))
    for x_batch, y_batch in train_generator:
        gc.collect()
        gradient = get_gradients(model, [x_batch, np.ones(y_batch.shape[0]), y_batch, 0], layers_name)
        mem = get_mem_usage()
        K.clear_session()
        K.set_session(tf.Session(graph))
        K.get_session().run(trainable_weights)
        for k,v in gradient.iteritems():
            gradients[k] = v + gradients[k] * total
            gradients[k] = gradients[k] / (total + x_batch.shape[0])
        total += x_batch.shape[0]
        print total, mem
        if total >= train_generator.n:
            break`

I want to get the gradient of image dataset (Imagenet) wrt total loss of vgg16 model on tensorflow backend, and this code runs correctly, but as the image number of dataset is large, so I use generator to calculate gradient and the batch is 32. However, the memory is increasing after each iteration, after hundrands of iteration, the memory is out of the limit, so the process is killed. I have used the method of K.clear_session, but as the graph is reset, it cannot be used in next iteration.
I don't know how to solve it, and get the gradient status of whole dataset.

**Comment by hermanc1** [comment]
I have the same problem. Has any progress been made?
